pkg/controller/jobframework/webhook/defaulter.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+
+	jsonpatch "gomodules.xyz/jsonpatch/v2"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// WithLosslessDefaulter creates a new Handler for a CustomDefaulter interface that **drops** remove operations,
+// which are typically the result of new API fields not present in Kueue libraries.
+func WithLosslessDefaulter(scheme *runtime.Scheme, obj runtime.Object, defaulter admission.CustomDefaulter) admission.Handler {
+	return &losslessDefaulter{
+		Handler: admission.WithCustomDefaulter(scheme, obj, defaulter).Handler,
+	}
+}
+
+type losslessDefaulter struct {
+	admission.Handler
+}
+
+// Handle handles admission requests, **dropping** remove operations from patches produced by controller-runtime.
+// The controller-runtime handler works by creating a jsondiff from the raw object and the marshalled
+// version of the object modified by the defaulter. This generates "remove" operations for fields
+// that are not present in the go types, which can occur when Kueue libraries are behind the latest
+// released CRDs.
+// Dropping the "remove" operations is safe because Kueue's job mutators never remove fields.
+func (h *losslessDefaulter) Handle(ctx context.Context, req admission.Request) admission.Response {
```

Comment:
I was chatting with @jpbetz to get his opinion on my former approach, doing a round-trip of the raw object, which was rejected by the maintainers of controller-runtime because it might break backwards compatibility when a go type removes an omitempty clause that was set for a field before https://github.com/kubernetes-sigs/controller-runtime/pull/2931#discussion_r1728367254

He agreed that both approaches (current behavior of controller-runtime and my PR) have pros and cons. However, getting new fields dropped is worse than dealing with CRDs that did backward-incompatible changes related to omitempty.
And, in the case of Kueue, we are only dealing with production ready CRDs and we only modify a few fields that the likelihood of a problem is low.
So, we could move in that direction instead of the approach in the current PR to drop "remove" operations, which can be less future proof.
WDYT?

Reply:
I support moving forward with https://github.com/kubernetes-sigs/controller-runtime/pull/2932 to at least enable us in Kueue to register the lossLesshandler. 

However, for the long term solution LossLessHandler isn't ideal. Even if we don't the problem yet, it might happen with increased adoption, for example an adopter of Kueue wants to write an in-house webhook to drop a problematic field from Volcano CRD (for example to work-around some issue). 

I see the concern in https://github.com/kubernetes-sigs/controller-runtime/pull/2931, but some solution like this could be handy, maybe it just needs additional care for the problematic cases mentioned?

I'm also wondering if this problem is related to https://github.com/kubernetes/kubernetes/issues/126175 (might be red herring, but PTAL)

Reply:
> but some solution like this could be handy, maybe it just needs additional care for the problematic cases mentioned?

We wouldn't be able to know generically. It depends on the CRD. But again, the likelihood is very low.

So, up to you. Would you prefer the implementation of LosslessHandler as suggested in https://github.com/kubernetes-sigs/controller-runtime/pull/2931 or are you ok with proceeding with the current PR?

Reply:
I will need to learn more about the pros & cons of thttps://github.com/kubernetes-sigs/controller-runtime/pull/2931. 
Would the change in omitempty be problematic then only for build-in types, or CRDs also? Also is it problematic removing "omitempty", or adding as well?

Reply:
The problematic scenario is when the go library in Kueue doesn't have omitempty, but the object stored in apiserver doesn't have the field. In this case, roundtripping in Kueue would add the fields, leading to possibly invalid patches. This can happen in two scenarios:
- The go type author removes the omitempty from a field, and this is the library that Kueue uses. A user created the object when omitempty was in place.
- Kueue is using an older library that doesn't have omitempty. The go type author adds omitempty to a field, and a user creates an object using this version.
So yes, it can happen in either scenario. @jpbetz, please correct me if my interpretation is wrong.

Another diminishing factor is that this only matters if we are modifying a field that is having these omitempty transitions. If the transition happens in an object that Kueue doesn't touch, then this wouldn't show up in the jsondiff. But the omitempty status of all the parents of the field could be relevant.

I think https://github.com/kubernetes/kubernetes/issues/126175 is also related, but it rather happens with the current CustomDefaulter of controller-runtime, and it wouldn't happen with my version because it wouldn't show in the jsondiff.

So the question comes down to:
Do we want to prioritize supporting newly added fields that Kueue libraries don't know about? Or do we prioritize supporting buggy types (which, as we can see, could also be a problem in the current version)? I think it's only fair to prioritize the former and be in the watch to report issues as necessary in other libraries and upgrade when appropriate.

OTOH, the LosslessDefaulter as currently implemented (just dropping every remove operation), doesn't have any problems, as long as we know that we are not removing any fields in the webhook.

Reply:
I think there is a risk in kubeflow because some of the `RunPolicy` fields don't have omitempty. Maybe we can make sure they get omitempty before we implement the RoundtripDefaulter in Kueue? cc @andreyvelich (sorry, it's a long thread to read, but you can only understand it if you follow all of it).

Reply:
I'm strongly in favor of biasing toward approaches that work for well-behaved CRD authors.

Adding fields to CRDs is very common, and it's essential to being able to safely and backward compatibly evolving a CRD over time.  If we have a webhook mechanism that removes fields that it doesn't recognize or need to care about purely because the webhook was developed before the field was introduced, that's a risk to all well behaved CRD authors. I personally wouldn't ever use a webhook that might start removing perfectly valid fields in the future.  That seems broken to me, and the failure mode is pretty bad.. it seems to work when you develop it, then later it mutates resources in a dangerous way, maybe in production.

Am I mis-representing the problem?

> Would the change in omitempty be problematic then only for build-in types, or CRDs also? Also is it problematic removing "omitempty", or adding as well?

I don't think we could ever make this change to a built-in type. And even if you do make this change, it's it only problem if the server cares about the difference between absent and zero-valued? (which is backward incompatible and dangerous for the interpretation of any resources written to storage before the change?)

Reply:
> Am I mis-representing the problem?

No, you are on point :)

> I don't think we could ever make this change to a built-in type. And even if you do make this change, isn't it only problem if the server differentiates between absent and zero-valued? (

Right, in a built-in type, we would never remove an "omitempty" for an optional field, but we could still add one, just because we forgot before, like in https://github.com/kubernetes/kubernetes/pull/126046

I think, in that case, if Kueue keeps the library with the buggy go type (without omitempty), then using the RoundTripperDefaulter is at risk of producing invalid jsonpatches.

Reply:
But in that case, it would be silly not to upgrade Kueue to the latest patch version.

Reply:
My overall suggestion is the following:
1. Merge LosslessDefaulter (this PR)
2. Do an audit of all in-tree integrations and make sure they are using omitempty appropriately, at least for the fields we modify.
3. Fix each external CRD as necessary and wait for their releases.
4. Upgrade Kueue libraries and change the LosslessDefaulter for a RoundTripDefaulter.

Reply:
Aldo, big thanks for explaining the issue in detail, and to all folks sharing their perspectives.

I fully agree with the need to fix the issue, a webhook which removes fields is very bad. This is important to the extent that I would consider cherry-picking this fix to 0.8.2. Even if the current master webhook doesn’t drop fields, because it is compiled against the latest CRD (say Ray), it may start dropping fields in a couple on months just because the next version of Ray is released (outside of Kueue’s control). WDY @tenzen-y ?

Regarding the fix of choice. 

Re 1. I really like the LossLessDefaulter approach, and I don’t see it buggy at all. IIUC it ensures correctness (current and future) for all CRDs managed by Kueue, as long as our the webhooks don’t remove fields, but our webhooks are part of Kueue codebase for built-in integrations, so we are in control. The approach is opt-in, so external developers may or may not want to use LossLessDefaulter (we should warn about the potential risk in documentation). Finally, if a user developing external or internal integration using LossLessDefaulter has a need to remove a field, they could do so using a second webhook registered using regular Defaulter, and just drop the field (set to nil). It will be safe, because the dedicated webhook will be compiled against the CRD version new enough to contain the field.

Re 2-4: I’m sceptical here. First, IIUC the mechanism as prototyped in https://github.com/kubernetes-sigs/controller-runtime/pull/2931 is not opt-in, and so if there is a field that changes omitempty, the correctness is out of Kueue - we basically need to accept the fact that Kueue’s webhooks misbehave from the end-user perspective.  

I just did a quick heuristic check: searched in kubernetes by “omitempty” : https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+omitempty+label%3Akind%2Fapi-change+is%3Aclosed , and clicked through a couple of top results, finding:
- https://github.com/kubernetes/kubernetes/pull/126046 added omitempty
- https://github.com/kubernetes/kubernetes/pull/124859 removed omitempty
- https://github.com/kubernetes/kubernetes/pull/124553 removed omitempty
- https://github.com/kubernetes/kubernetes/pull/121078 added omitempty
- https://github.com/kubernetes/kubernetes/pull/121000 added omitempty

I also remember OnCalling an issue where the culprit was that a user changed omitempty for an in-house CRD.

Given that such changes happen in the core k8s (often to fix bugs), and we cannot control in-house integrations, I think it isn't feasible to audit all CRDs (as suggested in 2-3 points), so I think any fix we choose needs to handle that scenario gracefully. 

I would like to move forward with (1.), and maybe (4) provided that the approach can be adjusted to handle changes in omitempty gracefully. 

I’m happy to re-evaluate if I missed something.

Reply:
One more thing to clarify: when I say I don't see (1.) buggy I mean it solves the problem for Kueue and as long as we are careful - the correctness is in hands of Kueue developers (or developers of integrations). **as long as we are careful** is still an important caveat here, so the solution is far from ideal. Another reason it is not ideal, is that other projects, like JobSet would also need to introduce something like LossLessDefaulter, and this approach will not scale.  It is just my preference out of the options on the table.

Reply:
> First, IIUC the mechanism as prototyped in https://github.com/kubernetes-sigs/controller-runtime/pull/2931 is not opt-in

Right, that original PR is not opt-in. However, the proposal would be to add a `RoundTripDefaulter` as an alternative to c-r's `CustomDefaulter`, not as a replacement. Then it is opt-in.

But, in any case, if you are accepting `LosslessDefaulter` as is, then the only reason to change to `RoundTripDefaulter` would be because we suddenly have a need to remove fields in our webhooks or we just want to be future-proof or provide the alternative to external developers.

A good thing about making it opt-in, is that we don't have to migrate all webhooks at once. We can do the audit for a specific type and then do the migration for just that webhook.

> I just did a quick heuristic check: searched in kubernetes by “omitempty” 

That's scary. We need to be more diligent about https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required when doing API reviews. I'll send an email to api-reviewers :)

> Another reason it is not ideal, is that other projects, like JobSet would also need to introduce something like LossLessDefaulter

I don't think this is a problem for Jobset, at least not in the same way. Jobset's webhook would always be released alongside the new CRD version. Sure, jobset uses job specs underneath, which in theory could add fields, but unless they are in the latest version, the fields would be dropped even if there wasn't a webhook, because they wouldn't be part of the jobset CRD.
So, this is more of a problem of components like Kueue that deal with libraries that it doesn't control.

Anyways, it sounds like we should be proceeding with this PR for now. Any additional implementation comments?

Reply:
> the proposal would be to add a RoundTripDefaulter as an alternative to c-r's CustomDefaulter, not as a replacement. Then it is opt-in.

Probably worth adding to controller-runtime as an option. However, even then I would be a bit cautious for using in Kueue (but open). The reason gets correctness, from the end-user perspective, out of Kueue. So, a user of Kueue may see there was an edit of CRD by Kueue's webhook, and attribute this as a bug of the Kueue webhook, which would be understandable, even though the actual change was in CRD.

> I don't think this is a problem for Jobset, at least not in the same way. Jobset's webhook would always be released alongside the new CRD version.

Ah, right. Though, to some extent it may still happen if omitempty is changed for the Pod API in Kube, because JobSet has [pod webook](https://github.com/kubernetes-sigs/jobset/blob/main/pkg/webhooks/pod_mutating_webhook.go) also.

> Anyways, it sounds like we should be proceeding with this PR for now. Any additional implementation comments?

Good to go from me, but let's see if @tenzen-y has some final comments.

Reply:
> I think there is a risk in kubeflow because some of the RunPolicy fields don't have omitempty. Maybe we can make sure they get omitempty before we implement the RoundtripDefaulter in Kueue?

On behalf of Kubeflow Training TL, I believe this request is reasonable and acceptable. But, I want to open an issue in the kubeflow side before we start implementations.

Another concern is the older Kubeflow TrainingOperator users. IIUC, after the RoundTripDefaulter is implemented in the Kueue side, if the users use the older TrainingOperator (without omitempty RunPolicy), some field could be accidentally dropped during webhook, right?

So, I want to keep using this LossLessDefaulter for a while. But, we can discuss it in another issue.

Reply:
>  if the users use the older TrainingOperator (without omitempty RunPolicy), some field could be accidentally dropped during webhook, right?

I don't think so. If Kueue libraries have omitempty when necessary, then the patches will be correct even if the objects were created with a different version.

Reply:
> > if the users use the older TrainingOperator (without omitempty RunPolicy), some field could be accidentally dropped during webhook, right?
> 
> I don't think so. If Kueue libraries have omitempty when necessary, then the patches will be correct even if the objects were created with a different version.

I see. I was misunderstaindin your point a little bit. Now, all clear. So, we want to add the `omitempty` to RunPolicy as soon as possible. Let's open an issue and add omitempty to the TrainigOperator and MPIOperator.

Thanks for the clarifications and for sharing your investigations.

Reply:
I think understanding all nuances of setting "omitempty" properly has proved to be hard, and so it is hard for me to imagine we could maintain bug-free CRDs. Even if we put effort into auditing the CRDs for version X, I think with new fields added such bugs are likely to slip review in X+2, and will require changes (bugfixes). 

I'm also a bit concerned that with widespread use of RoundTripDefauler we will have an incentive not to fix omitempty bugs, so that we don't run into issues on upgrades.

Reply:
> I'm also a bit concerned that with widespread use of RoundTripDefauler we will have an incentive not to fix omitempty bugs, so that we don't run into issues on upgrades.

I'm sure your concerns. But, I guess that we need to take a little bit longer time to fix the root cause in the contoller-runtime and core kube API. So, I would recommend to treat the RoutTripDefaulter as a temporary solution and put the function in the internal package so that the function is not widespread used.

Reply:
Again, the omitempty issue only affects Kueue webhooks if they are missing in the fields that Kueue modifies. If Kueue doesn't modify them, then they would simply not form part of the jsondiff. So, to ensure Kueue correctness, we only need to audit the fields that we touch.

Reply:
I guess that Michal's concern is for the third-party Kueue controllers. Isn't it?

Reply:
Yeah, but unfortunately we cannot do much better than document the problem.

Reply:
My comment was more speculative, touching two potential concerns. Let me explain below in more detail.

> So, to ensure Kueue correctness, we only need to audit the fields that we touch. 

Yes, but even if we audit all fields modified by Kueue for TFJob at a given release of Kueue and TFJob, Kueue may want to use new fields in its future releases (for new features, like managedBy). Then, we need to audit the new field, and if we find an "omitempty" bug in TFJob operator in the field, then what do we do? Wait with the new feature of Kueue until the TFJob is fixed and released, or build the new feature based on the buggy field? None of the choices is ideal.

> I guess that Michal's concern is for the third-party Kueue controllers. Isn't it?

Yeah, this is another consideration, if RoundTrip defaulter is widespread in customer, and third-party webhooks, it would create an incentive not to fix the omitempty bugs (whatever small, but still).